File: lib/rs/core/src/models/block_content.rs
use crate::models::NuttyTag;
use serde::{Deserialize, Serialize};

/// Not to be confused with [ContentBlock].
/// `ContentBlockContent` it might have been named,
/// but `BlockContent` is shorter and unclaimed.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum BlockContent {
	Page { title: String },
	Heading { markdown: String },
	Paragraph { markdown: String },
}

impl BlockContent {
	/// Parse the target [NuttyTag] list from the content block.
	pub fn parse_target_tags(&self) -> Vec<NuttyTag> {
		match self {
			BlockContent::Page { .. } => vec![],
			BlockContent::Heading { markdown } => NuttyTag::parse_all(markdown),
			BlockContent::Paragraph { markdown } => NuttyTag::parse_all(markdown),
		}
	}
}

File: app/src/models/block-content.ts
import { Schema } from "effect";

/**
 * Schema for content block content.
 *
 * Not to be confused with ContentBlock.
 * `ContentBlockContent` it might have been named,
 * but `BlockContent` is shorter and unclaimed.
 */
const BlockContent = Schema.Union(
	Schema.Struct({
		kind: Schema.Literal("Page"),
		title: Schema.String,
	}),
	Schema.Struct({
		kind: Schema.Literal("Heading"),
		markdown: Schema.String,
	}),
	Schema.Struct({
		kind: Schema.Literal("Paragraph"),
		markdown: Schema.String,
	}),
).pipe(
	Schema.annotations({
		identifier: "BlockContent",
		title: "Block Content",
		description: "Content for a block in the Nuttyverse",
	}),
);

type BlockContent = typeof BlockContent.Type;

export { BlockContent };
